src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+    string = true,
+    bytes = true,
+    message = true,
+}
+
+local protobuf_I32_types = {
+    fixed32 = true,
+    sfixed32 = true,
+}
+
+local protobuf_I64_types = {
+    fixed64 = true,
+    sfixed64 = true,
+}
+
+local c_integer = {
+    ['ctype<int64_t>'] = true,
+    ['ctype<uint64_t>'] = true,
+}
+
+local input_varint_types = {
+    boolean = true,
+    number = true,
+    cdata = true,
+}
+
+local input_len_types = {
+    string = true,
+}
+
+local input_I32_types = {
+    number = true,
+    cdata = true,
+}
+
+local input_I64_types = {
+    number = true,
+    cdata = true
+}
+
+-- Create a message object suitable to pass
+-- into the protobuf.protocol function.
+--
+-- Accepts a name of the message and a message
+-- definition in the following format.
+--
+-- message_def = {
+--    <field_name> = {<field_type>, <field_id>},
+--    <...>
+-- }
+local function message(message_name, message_def)
+    local field_by_name = {}
+    local field_by_id = {}
+    for field_name, def in pairs(message_def) do
+        local field_type = def[1]
+        local field_id = def[2]
+        if field_by_id[field_id] ~= nil then
+            error(('Id %d in field %q was already used'):format(field_id,
+                field_name))
+        end
+        if field_id < min_field_id or field_id > max_field_id then
+            error(('Id %d in field %q is out of range [%d; %d]'):format(
+                field_id, field_name, min_field_id, max_field_id))
+        end
+        if field_id >= reserved_field_id_min and
+           field_id <= reserved_field_id_max then
+           error(('Id %d in field %q is in reserved ' ..
+               'id range [%d, %d]'):format(field_id, field_name,
+               reserved_field_id_min, reserved_field_id_max))
+        end
+        local field_def = {
+            type = field_type,
+            name = field_name,
+            id = field_id,
+        }
+        field_by_name[field_name] = field_def
+        field_by_id[field_id] = field_def
+    end
+    return {
+        type = 'message',
+        name = message_name,
+        field_by_name = field_by_name,
+        field_by_id = field_by_id
+    }
+end
+
+-- Create a enum object suitable to pass into
+-- the protobuf.protocol function.
+--
+-- Accepts a name of an enum and an enum definition
+-- in the following format.
+--
+-- enum_def = {
+--     <value_name> = <value_id>,
+--     <...>
+-- }
+local function enum(enum_name, enum_def)
+    local value_by_name = {}
+    local value_by_id = {}
+    for value_name, value_id in pairs(enum_def) do
+        if value_by_id[value_id] ~= nil then
+            error(('Double definition of enum field %q by %d'):format(
+                value_name, value_id))
+        end
+        value_by_name[value_name] = value_id
+        value_by_id[value_id] = value_name
+    end
+    return {
+        type = 'enum',
+        name = enum_name,
+        value_by_name = value_by_name,
+        value_by_id = value_by_id,
+    }
```

Comment:
Sorry for the confusion, but the `value_by_name` was considered as a misleading name and replaced with `id_by_value` in the RFC document. So, we have `id_by_value` and `value_by_id` now.

If you consider that `id_by_name` and `name_by_id` are better, I'm OK with it too (will update in the RFC document).

Reply:
Replaced with id_by_value according to RFC